core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,69 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+    def tryCreateControllerZNodeAndIncrementEpoch(): (Int, Int) = {
+      try {
+        val transaction = zooKeeperClient.createTransaction()
+        transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
+          acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
+        transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
+        val results = transaction.commit()
+        val setDataResult = results.get(1).asInstanceOf[SetDataResult]
+        (newControllerEpoch, setDataResult.getStat.getVersion)
+      } catch {
+        case _: NodeExistsException =>
+          val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+            s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+              s"Aborting controller startup procedure"))
+          if (controllerId == curControllerId) {
+            val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
+              "Controller moved to another broker. Aborting controller startup procedure"))
+            // If the epoch is the same as newControllerEpoch, it is safe to infer that the returned epoch zkVersion
+            // is associated with the current broker during controller election because we already knew that the zk
+            // transaction succeeds based on the controller znode verification. Other rounds of controller
+            // election will result in larger epoch number written in zk.
+            if (epoch == newControllerEpoch)
+              return (newControllerEpoch, stat.getVersion)
+          }
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case _: BadVersionException =>
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case _: ConnectionLossException => tryCreateControllerZNodeAndIncrementEpoch()
+      }
+    }
+
+    tryCreateControllerZNodeAndIncrementEpoch()
+  }
+
+  private def createControllerEpochZNode(): (Int, Int) = {
+    createControllerEpochRaw(KafkaController.InitialControllerEpoch - 1).resultCode match {
+      case Code.OK =>
+        info(s"Successfully create ${ControllerEpochZNode.path} with initial epoch ${KafkaController.InitialControllerEpoch - 1}")
+        (KafkaController.InitialControllerEpoch - 1, KafkaController.InitialControllerEpochZkVersion - 1)
+      case Code.NODEEXISTS =>
+        throw new ControllerMovedException(s"Controller moved to another broker while creating ${ControllerEpochZNode.path}. Aborting controller startup procedure")
```

Comment:
It is possible that the controller_epoch path is created by another broker between getControllerEpoch() and createControllerEpochZNode(). In this case, we want to get the controller epoch again, instead of throwing ControllerMovedException.

Reply:
If the controller_epoch path is created by another broker between getControllerEpoch() and createControllerEpochZNode(), I was thinking whether we can infer that other broker wins in this round of controller election even if it hasn't created the controller znode. 

After a second thought, I think we should follow what you suggested for extra safety because if the broker fails to talk to zk for some reason, the cluster will get into a no-controller state.

Reply:
Fixed.